File: mycollab-scheduler/src/main/java/com/esofthead/mycollab/schedule/email/project/impl/MessageRelayEmailNotificationActionImpl.java
/**
 * This file is part of mycollab-scheduler.
 *
 * mycollab-scheduler is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mycollab-scheduler is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mycollab-scheduler.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.esofthead.mycollab.schedule.email.project.impl;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.config.BeanDefinition;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Service;

import com.esofthead.mycollab.common.domain.SimpleRelayEmailNotification;
import com.esofthead.mycollab.core.utils.StringUtils;
import com.esofthead.mycollab.module.mail.TemplateGenerator;
import com.esofthead.mycollab.module.project.ProjectLinkGenerator;
import com.esofthead.mycollab.module.project.domain.SimpleMessage;
import com.esofthead.mycollab.module.project.i18n.MessageI18nEnum;
import com.esofthead.mycollab.module.project.service.MessageService;
import com.esofthead.mycollab.module.project.service.ProjectService;
import com.esofthead.mycollab.schedule.email.MailContext;
import com.esofthead.mycollab.schedule.email.project.MessageRelayEmailNotificationAction;

/**
 * 
 * @author MyCollab Ltd.
 * @since 1.0
 * 
 */
@Service
@Scope(BeanDefinition.SCOPE_PROTOTYPE)
public class MessageRelayEmailNotificationActionImpl extends
		SendMailToAllMembersAction<SimpleMessage> implements
		MessageRelayEmailNotificationAction {

	@Autowired
	private MessageService messageService;
	@Autowired
	private ProjectService projectService;

	@Override
	protected TemplateGenerator templateGeneratorForCreateAction(
			MailContext<SimpleMessage> context) {
		SimpleMessage message = messageService.findMessageById(
				context.getTypeid(), context.getSaccountid());

		if (message == null) {
			return null;
		}

		context.setWrappedBean(message);
		TemplateGenerator templateGenerator = new TemplateGenerator(
				context.getMessage(MessageI18nEnum.MAIL_CREATE_ITEM_SUBJECT,
						message.getProjectName(),
						context.getChangeByUserFullName(),
						StringUtils.trim(message.getTitle(), 100)),
				context.templatePath("templates/email/project/itemCreatedNotifier.mt"));

		buildExtraTemplateVariables(context.getEmailNotification());

		templateGenerator.putVariable("message", message.getMessage());
		return templateGenerator;
	}

	@Override
	protected String getItemName() {
		return StringUtils.trim(bean.getTitle(), 100);
	}

	@Override
	protected String getCreateSubject(MailContext<SimpleMessage> context) {
		return context.getMessage(MessageI18nEnum.MAIL_CREATE_ITEM_SUBJECT,
				bean.getProjectName(), context.getChangeByUserFullName(),
				getItemName());
	}

	@Override
	protected TemplateGenerator templateGeneratorForUpdateAction(
			MailContext<SimpleMessage> context) {
		SimpleMessage message = messageService.findMessageById(
				context.getTypeid(), context.getSaccountid());

		if (message == null) {
			return null;
		}
		context.setWrappedBean(message);
		TemplateGenerator templateGenerator = new TemplateGenerator(
				context.getMessage(MessageI18nEnum.MAIL_UPDATE_ITEM_SUBJECT,
						message.getProjectName(),
						context.getChangeByUserFullName(),
						StringUtils.trim(message.getTitle(), 100)),
				context.templatePath("templates/email/project/itemCreatedNotifier.mt"));
		buildExtraTemplateVariables(context.getEmailNotification());

		templateGenerator.putVariable("message", message.getMessage());
		return templateGenerator;
	}

	@Override
	protected TemplateGenerator templateGeneratorForCommentAction(
			MailContext<SimpleMessage> context) {
		SimpleMessage message = messageService.findMessageById(
				context.getTypeid(), context.getSaccountid());

		if (message == null) {
			return null;
		}

		TemplateGenerator templateGenerator = new TemplateGenerator(
				context.getMessage(MessageI18nEnum.MAIL_COMMENT_ITEM_SUBJECT,
						message.getProjectName(),
						context.getChangeByUserFullName(),
						StringUtils.trim(message.getTitle(), 100)),
				context.templatePath("templates/email/project/itemCommentNotifier.mt"));
		buildExtraTemplateVariables(context.getEmailNotification());

		templateGenerator
				.putVariable("comment", context.getEmailNotification());

		return templateGenerator;
	}

	@Override
	protected SimpleMessage getBeanInContext(MailContext<SimpleMessage> context) {
		return messageService.findMessageById(context.getTypeid(),
				context.getSaccountid());
	}

	@Override
	protected void buildExtraTemplateVariables(
			SimpleRelayEmailNotification emailNotification) {
		List<Map<String, String>> listOfTitles = new ArrayList<Map<String, String>>();

		HashMap<String, String> currentProject = new HashMap<String, String>();
		currentProject.put("displayName", bean.getProjectName());
		currentProject.put(
				"webLink",
				ProjectLinkGenerator.generateProjectFullLink(siteUrl,
						bean.getProjectid()));

		listOfTitles.add(currentProject);

		String summary = bean.getTitle();
		String summaryLink = ProjectLinkGenerator
				.generateMessagePreviewFullLink(siteUrl, bean.getProjectid(),
						bean.getId());

		contentGenerator.putVariable("makeChangeUser",
				emailNotification.getChangeByUserFullName());
		contentGenerator.putVariable("itemType", "message");
		contentGenerator.putVariable("titles", listOfTitles);
		contentGenerator.putVariable("summary", summary);
		contentGenerator.putVariable("summaryLink", summaryLink);

	}
}

File: mycollab-scheduler/src/main/scala/com/esofthead/mycollab/schedule/email/project/impl/MessageRelayEmailNotificationActionImpl.scala
package com.esofthead.mycollab.schedule.email.project.impl

import com.esofthead.mycollab.common.MonitorTypeConstants
import com.esofthead.mycollab.common.domain.SimpleRelayEmailNotification
import com.esofthead.mycollab.configuration.StorageManager
import com.esofthead.mycollab.core.utils.StringUtils
import com.esofthead.mycollab.module.project.ProjectLinkGenerator
import com.esofthead.mycollab.module.project.domain.{SimpleMessage, SimpleProjectMember}
import com.esofthead.mycollab.module.project.i18n.MessageI18nEnum
import com.esofthead.mycollab.module.project.service.{MessageService, ProjectMemberService, ProjectService}
import com.esofthead.mycollab.schedule.email.project.MessageRelayEmailNotificationAction
import com.esofthead.mycollab.schedule.email.{ItemFieldMapper, MailContext}
import com.hp.gagawa.java.elements.Img
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.config.BeanDefinition
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Service

/**
 * @author MyCollab Ltd.
 * @since 4.6.0
 */
@Service
@Scope(BeanDefinition.SCOPE_PROTOTYPE)
class MessageRelayEmailNotificationActionImpl extends
SendMailToAllMembersAction[SimpleMessage] with MessageRelayEmailNotificationAction {

  @Autowired var messageService: MessageService = _

  @Autowired var projectService: ProjectService = _

  @Autowired var projectMemberService: ProjectMemberService = _

  protected def getItemName: String = StringUtils.trim(bean.getTitle, 100)

  protected def getCreateSubject(context: MailContext[SimpleMessage]): String = context.getMessage(MessageI18nEnum.MAIL_CREATE_ITEM_SUBJECT, bean.getProjectName, context.getChangeByUserFullName, getItemName)

  protected def getUpdateSubject(context: MailContext[SimpleMessage]): String = context.getMessage(MessageI18nEnum.MAIL_UPDATE_ITEM_SUBJECT, bean.getProjectName, context.getChangeByUserFullName, getItemName)

  protected def getCommentSubject(context: MailContext[SimpleMessage]): String = context.getMessage(MessageI18nEnum.MAIL_COMMENT_ITEM_SUBJECT, bean.getProjectName, context.getChangeByUserFullName, getItemName)

  protected def getItemFieldMapper: ItemFieldMapper = null

  protected def getBeanInContext(context: MailContext[SimpleMessage]): SimpleMessage = messageService.findById(context.getTypeid.toInt, context.getSaccountid)

  protected def buildExtraTemplateVariables(context: MailContext[SimpleMessage]) {
    val listOfTitles: List[Map[String, String]] = List[Map[String, String]]()
    val emailNotification: SimpleRelayEmailNotification = context.getEmailNotification
    val currentProject: Map[String, String] = Map[String, String]()
    currentProject.+("displayName" -> bean.getProjectName)
    currentProject.+("webLink" -> ProjectLinkGenerator.generateProjectFullLink(siteUrl, bean.getProjectid))
    listOfTitles.+:(currentProject)
    val summary: String = bean.getTitle
    val summaryLink: String = ProjectLinkGenerator.generateMessagePreviewFullLink(siteUrl, bean.getProjectid, bean.getId)
    var avatarId: String = ""
    val projectMember: SimpleProjectMember = projectMemberService.findMemberByUsername(emailNotification.getChangeby, bean.getProjectid, emailNotification.getSaccountid)
    if (projectMember != null) {
      avatarId = projectMember.getMemberAvatarId
    }
    val userAvatar: Img = new Img("", StorageManager.getAvatarLink(avatarId, 16))
    userAvatar.setWidth("16")
    userAvatar.setHeight("16")
    userAvatar.setStyle("display: inline-block; vertical-align: top;")
    val makeChangeUser: String = userAvatar.toString + emailNotification.getChangeByUserFullName
    if (MonitorTypeConstants.CREATE_ACTION == emailNotification.getAction) {
      contentGenerator.putVariable("actionHeading", context.getMessage(MessageI18nEnum.MAIL_CREATE_ITEM_HEADING, makeChangeUser))
    }
    else if (MonitorTypeConstants.UPDATE_ACTION == emailNotification.getAction) {
      contentGenerator.putVariable("actionHeading", context.getMessage(MessageI18nEnum.MAIL_UPDATE_ITEM_HEADING, makeChangeUser))
    }
    else if (MonitorTypeConstants.ADD_COMMENT_ACTION == emailNotification.getAction) {
      contentGenerator.putVariable("actionHeading", context.getMessage(MessageI18nEnum.MAIL_COMMENT_ITEM_HEADING, makeChangeUser))
    }
    contentGenerator.putVariable("titles", listOfTitles)
    contentGenerator.putVariable("summary", summary)
    contentGenerator.putVariable("summaryLink", summaryLink)
    contentGenerator.putVariable("message", bean.getMessage)
  }
}
